privacyidea/api/lib/prepolicy.py

```diff
@@ -1422,3 +1431,447 @@ def allowed_audit_realm(request=None, action=None):
+            = challengetext
+
+        request.all_data["scope"] = SCOPE.AUTH
+
+    return True
+
+
+def webauthntoken_enroll(request, action):
+    """
+    This is a token specific wrapper for the WebAuthn token for the endpoint /token/init.
+
+    This will enrich the initialization request for WebAuthn tokens with the
+    necessary configuration information from policy actions with
+    scope=SCOPE.ENROLL. The request will be augmented with a name and id for
+    the relying party, as specified by the with actions
+    WEBAUTHNACTION.RELYING_PARTY_NAME and WEBAUTHNACTION.RELYING_PARTY_ID,
+    respectively, authenticator attachment preference, public key credential
+    algorithm preferences, authenticator attestation requirement level,
+    authenticator attestation requirement form, allowed AAGUIDs, and the text
+    to display to the user when asking to confirm the challenge on the token,
+    as specified by the actions WEBAUTHNACTION.AUTHENTICATOR_ATTACHMENT,
+    WEBAUTHNACTION.PUBLIC_KEY_CREDENTIAL_ALGORITHM_PREFERENCE,
+    WEBAUTHNACTION.AUTHENTICATOR_ATTESTATION_LEVEL,
+    WEBAUTHNACTION.AUTHENTICATOR_ATTESTATION_FORM,
+    WEBAUTHNACTION.AUTHENTICATOR_SELECTION_LIST, and
+    ACTION.CHALLENGETEXT, respectively.
+
+    Setting WEBAUTHNACTION.RELYING_PARTY_NAME and
+    WEBAUTHNACTION.RELYING_PARTY_ID is mandatory, and if either of these is not
+    set, we bail out.
+
+    This will also add a parameter called `scope` to the request, with a value
+    of SCOPE.ENROLL. This is later consumed by the webauthn_request() decodator.
+
+    :param request:
+    :type request:
+    :param action:
+    :type action:
+    :return:
+    :rtype:
+    """
+
+    ttype = request.all_data.get("type")
+    if ttype and ttype.lower() == WebAuthnTokenClass.get_class_type():
+        rp_id_policies = Match\
+            .user(g,
+                  scope=SCOPE.ENROLL,
+                  action=WEBAUTHNACTION.RELYING_PARTY_ID,
+                  user_object=request.User if request.User else None)\
+            .action_values(unique=True)
+        if rp_id_policies:
+            rp_id = list(rp_id_policies)[0]
+        else:
+            raise PolicyError("Missing enrollment policy for WebauthnToken: " + WEBAUTHNACTION.RELYING_PARTY_ID)
+
+        rp_name_policies = Match\
+            .user(g,
+                  scope=SCOPE.ENROLL,
+                  action=WEBAUTHNACTION.RELYING_PARTY_NAME,
+                  user_object=request.User if request.User else None)\
+            .action_values(unique=True,
+                           allow_white_space_in_action=True)
+        if rp_name_policies:
+            rp_name = list(rp_name_policies)[0]
+        else:
+            raise PolicyError("Missing enrollment policy for WebauthnToken: " + WEBAUTHNACTION.RELYING_PARTY_NAME)
+
+        # The RP ID is a domain name and thus may not contain any punctuation except '-'.
+        if not set(string.punctuation).intersection(rp_id).issubset(set('-')):
```

Comment:
Interesting, isn't there a more straightforward way to test for hostnames? 
If not, maybe we should have this in our own utils.py.

Reply:
Please move it to a utils: check_fqdn()